parsers/US_SPP.py

```diff
@@ -0,0 +1,198 @@
+#!usr/bin/env python3
+
+"""Parser for the Southwest Power Pool area of the United States."""
+
+from dateutil import parser, tz
+from io import StringIO
+from logging import getLogger
+import pandas as pd
+import requests
+
+generation_url = 'https://marketplace.spp.org/public-data-api/gen-mix/asFile'
+
+exchange_url = 'https://marketplace.spp.org/public-data-api/interchange-trend/asFile'
+
+mapping = {'coal': 'coal',
+           'unknown': 'unknown',
+           'Wind': 'wind',
+           'Nuclear': 'nuclear',
+           'Hydro': 'hydro',
+           'Solar': 'solar',
+           'Natural Gas': 'gas',
+           'Diesel Fuel Oil': 'oil',
+           'Waste Disposal Services': 'biomass'
+            }
+
+tie_mapping = {'US-MISO->US-SPP': ['AMRN', 'DPC', 'GRE', 'MDU', 'MEC', 'NSP', 'OTP']}
+
+# NOTE
+# Data sources return timestamps in GMT.
+# Energy storage situation unclear as of 16/03/2018, likely to change quickly in future.
+
+
+def get_data(url, session=None):
+    """Returns a pandas dataframe."""
+
+    s=session or requests.Session()
+    req = s.get(url, verify=False)
+    df = pd.read_csv(StringIO(req.text))
+
+    return df
+
+
+def data_processor(df, logger):
+    """
+    Takes a dataframe and logging instance as input.
+    Checks for new generation types and logs awarning if any are found.
+    Parses the dataframe row by row removing unneeded keys.
+    Returns a list of 2 element tuples, each containing a datetime object
+    and production dictionary.
+    """
+
+    # Remove leading whitespace in column headers.
+    df.columns = df.columns.str.strip()
+
+    keys_to_remove = ['Coal Market', 'Coal Self', 'GMT MKT Interval', 'Average Actual Load',
+                      'Other', 'Waste Heat']
+
+    # Check for new generation columns.
+    known_keys = list(mapping.keys()) + keys_to_remove
+    column_headers = list(df)
+
+    unknown_keys = []
+    for heading in column_headers:
+        if heading not in known_keys:
+            logger.warning('New column \'{}\' present in US-SPP data source.'.format(
+                heading), extra={'key': 'US-SPP'})
+            unknown_keys.append(heading)
+
+    keys_to_remove += unknown_keys
```

Comment:
Completely optional, the above switched to using sets:

```
keys_to_remove = {'Coal Market', 'Coal Self', 'GMT MKT Interval', 'Average Actual Load',
    'Other', 'Waste Heat'}  # set literal
known_keys = mapping.keys() | keys_to_remove  # this is a set union
column_headers = set(df.columns)  # check exact syntax

unknown_keys = column_headers - known_keys  # set difference

for heading in unknown_keys:
    logger.warning('New column \'{}\' present in US-SPP data source.'.format(heading),
        extra={'key': 'US-SPP'})

keys_to_remove = keys_to_remove | unknown_keys  # can also do "|=" but it's a bit unclear I feel
```

Untested! But later on in the code, sets can be iterated over same as lists.

Reply:
Would you say there is a benefit to using sets here rather than lists?

Reply:
it's certainly not necessary, but makes computing `unknown_keys` easier and clearer